kernel/cooking.ml

```diff
@@ -185,32 +185,37 @@ let cook_constr { Opaqueproof.modlist ; abstract = (vars, subst, _) } c =
   let hyps = abstract_context hyps in
   abstract_constant_body (expmod c) hyps
 
+let discharge_abstract_universe_context subst abs_ctx auctx =
+  (** Given a named instance [subst := u₀ ... uₙ₋₁] together with an abstract
+      context [auctx0 := 0 ... n - 1 |= C{0, ..., n - 1}] of the same length,
+      and another abstract context relative to the former context
+      [auctx := 0 ... m - 1 |= C'{u₀, ..., uₙ₋₁, 0, ..., m - 1}],
+      construct the lifted abstract universe context
+      [0 ... n - 1 n ... n + m - 1 |=
+        C{0, ... n - 1} ∪
+        C'{0, ..., n - 1, n, ..., n + m - 1} ]
+      together with the instance
+      [u₀ ... uₙ₋₁ Var(0) ... Var (m - 1)].
+  *)
+  if (Univ.Instance.is_empty subst) then
+    (** Still need to take the union for the constraints between globals *)
+    subst, (AUContext.union abs_ctx auctx)
+  else
+    let open Univ in
+    let ainst = make_abstract_instance auctx in
+    let subst = Instance.append subst ainst in
+    let substf = make_instance_subst subst in
+    let auctx = Univ.subst_univs_level_abstract_universe_context substf auctx in
+    subst, (AUContext.union abs_ctx auctx)
+
 let lift_univs cb subst auctx0 =
   match cb.const_universes with
   | Monomorphic ctx ->
     assert (AUContext.is_empty auctx0);
     subst, (Monomorphic ctx)
   | Polymorphic auctx ->
-    (** Given a named instance [subst := u₀ ... uₙ₋₁] together with an abstract
-        context [auctx0 := 0 ... n - 1 |= C{0, ..., n - 1}] of the same length,
-        and another abstract context relative to the former context
-        [auctx := 0 ... m - 1 |= C'{u₀, ..., uₙ₋₁, 0, ..., m - 1}],
-        construct the lifted abstract universe context
-        [0 ... n - 1 n ... n + m - 1 |=
-          C{0, ... n - 1} ∪
-          C'{0, ..., n - 1, n, ..., n + m - 1} ]
-        together with the instance
-        [u₀ ... uₙ₋₁ Var(0) ... Var (m - 1)].
-    *)
-    if (Univ.Instance.is_empty subst) then
-      (** Still need to take the union for the constraints between globals *)
-      subst, (Polymorphic (AUContext.union auctx0 auctx))
-    else
-      let ainst = Univ.make_abstract_instance auctx in
-      let subst = Instance.append subst ainst in
-      let substf = Univ.make_instance_subst subst in
-      let auctx' = Univ.subst_univs_level_abstract_universe_context substf auctx in
-      subst, (Polymorphic (AUContext.union auctx0 auctx'))
```

Comment:
I guess you see it as moving this block rather than duplicating the lib function